File: mother_vertex.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class Solution 
{
    vector<int> v;
    int count;
    void dfs(int k, vector<int> adj[], vector<int>& f) {
        v[k] = 1;
        for(int i = 0; i < adj[k].size(); ++i) {
            if (v[adj[k][i]] == 0) {
                dfs(adj[k][i], adj, f);
            }
        }
        f[k] = count++;
    }
    public:
    //Function to find a Mother Vertex in the Graph.
	int findMotherVertex(int V, vector<int>adj[])
	{
	    // Code here
	    vector<int> f(V);
	    v.assign(V, 0);
	    count = 0;
	    for(int i = 0; i < V; ++i) {
	        v.assign(V, 0);
	        count = 0;
	        dfs(i, adj, f);
	        if (f[i] == V-1) return i;
	    }
	    
	    return -1;
	}

};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int V, E;
		cin >> V >> E;
		vector<int>adj[V];
		for(int i = 0; i < E; i++){
			int u, v;
			cin >> u >> v;
			adj[u].push_back(v);
		}
		Solution obj;
		int ans = obj.findMotherVertex(V, adj);
		cout << ans <<"\n";
	}
	return 0;
}
// } Driver Code Ends

File: graph/mother_vertex.py
class Solution:
    # Function to find a Mother Vertex in the Graph.
    def findMotherVertex(self, V, adj):
        # Code here
        visited = [0] * V

        def dfs(v, s):
            visited[v] = 1
            s.add(v)
            for w in adj[v]:
                if w not in s:
                    dfs(w, s)

        for v in range(V):
            s = set()
            if visited[v] == 0:
                dfs(v, s)
            if len(s) == V:
                return v

        return -1


# {
# Driver Code Starts
# Initial Template for Python 3

import sys

sys.setrecursionlimit(10**6)
if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        V, E = map(int, input().split())
        adj = [[] for i in range(V)]
        for _ in range(E):
            u, v = map(int, input().split())
            adj[u].append(v)
        obj = Solution()
        ans = obj.findMotherVertex(V, adj)
        print(ans)
# } Driver Code Ends
